File: src/test/java/dev/kensa/sentence/DictionaryTest.java
package dev.kensa.sentence;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;

import java.util.List;

import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;


class DictionaryTest {

    private Dictionary dictionary;

    @BeforeEach
    void setUp() {
        dictionary = new Dictionary();
    }

    @AfterEach
    void tearDown() {
        dictionary.clearAcronyms();
    }

    @Test
    void canStreamExistingAcronyms() {
        Acronym[] acronyms = {
                Acronym.of("foo", "means foo"),
                Acronym.of("boo", "means boo"),
                Acronym.of("moo", "means moo")
        };

        dictionary.putAcronyms(acronyms);

        assertThat(dictionary.acronyms()).containsAll(asList(acronyms));
    }

    @Test
    void canPutKeywords() {
        dictionary.putKeyword("foo");
        dictionary.putKeywords("boo", "moo");

        assertThat(dictionary.keywords()).containsAll(List.of("foo", "boo", "moo"));
    }

    @Test
    void canClearAcronyms() {
        dictionary.putAcronyms(Acronym.of("foo", "foo"));

        dictionary.clearAcronyms();

        assertThat(dictionary.acronyms()).isEmpty();
    }

    @ParameterizedTest
    @NullAndEmptySource
    void throwsOnAttemptToAddEmptyOrNullKeyword(String keyword) {
        assertThatThrownBy(() -> dictionary.putKeyword(keyword))
                .isInstanceOf(IllegalArgumentException.class);

        assertThatThrownBy(() -> dictionary.putKeywords("foo", keyword))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void throwsOnAttemptToAddEmptyOrNullAcronym() {
        assertThatThrownBy(() -> dictionary.putAcronyms((Acronym[]) null))
                .isInstanceOf(NullPointerException.class);
    }
}

File: src/test/java/dev/kensa/sentence/DictionaryTest.kt
package dev.kensa.sentence

import dev.kensa.sentence.Acronym.Companion.of
import io.kotest.assertions.throwables.shouldThrowExactly
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldContainAll
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

internal class DictionaryTest {
    private lateinit var dictionary: Dictionary

    @BeforeEach
    fun setUp() {
        dictionary = Dictionary()
    }

    @AfterEach
    fun tearDown() {
        dictionary.clearAcronyms()
    }

    @Test
    fun canStreamExistingAcronyms() {
        val acronyms = arrayOf(
                of("foo", "means foo"),
                of("boo", "means boo"),
                of("moo", "means moo")
        )
        dictionary.putAcronyms(*acronyms)
        dictionary.acronyms shouldContainAll acronyms.toList()
    }

    @Test
    fun canPutKeywords() {
        dictionary.putKeyword("foo")
        dictionary.putKeywords("boo", "moo")
        dictionary.keywords shouldContainAll listOf("foo", "boo", "moo").map { Keyword(it) }
    }

    @Test
    fun canClearAcronyms() {
        dictionary.putAcronyms(of("foo", "foo"))
        dictionary.clearAcronyms()
        dictionary.acronyms.shouldBeEmpty()
    }

    @ParameterizedTest
    @MethodSource("arguments")
    fun throwsOnAttemptToAddInvalidKeyword(keyword: String) {
        shouldThrowExactly<IllegalArgumentException> { dictionary.putKeyword(keyword) }
        shouldThrowExactly<IllegalArgumentException> { dictionary.putKeywords("foo", keyword) }
    }

    companion object {
        @JvmStatic
        fun arguments(): Stream<Arguments> {
            return Stream.of(
                    Arguments.of(""),
                    Arguments.of("   "),
                    Arguments.of("x"),
                    Arguments.of("xx")
            )
        }
    }
}